Support defining workflows restriction for API keys

slobodanadamovic opened the pull request:
The changes in this PR extend `RoleDescriptor` to support 
specifying `workflows` `restriction` for API keys:

```
POST /_security/api_key
{
  "name": "my_restricted_api_key",
  "role_descriptors": {
    "my_restricted_role": {
      "indices": [
        {
          "names": ["books", "movies"],
          "privileges": ["read"]
        }
      ],
      "restriction": {
        "workflows": ["my_workflow"]
      }
    }
  }
}
```

The workflows will be defined statically and used to restrict 
the API key's access to only specific REST APIs.

To keep the scope of the PR small, I will address below items in follow up PRs:

- Workflow definition and resolution
- Validation logic to verify that a workflow name is valid
- Blocking workflows restriction to be specified via Roles API
- Blocking file-based definitions of roles with workflows restriction
- Enforcing of workflow restrictions for API keys

slobodanadamovic committed:
Allow defining workflows restriction for role descriptors

slobodanadamovic committed:
Resolve compilation errors.

slobodanadamovic committed:
Fix another compilation error. Add TODO.

slobodanadamovic committed:
Fix API key BWC test

slobodanadamovic committed:
Fix failing test

slobodanadamovic committed:
Test role descriptor parsing

slobodanadamovic committed:
Refactor restriction serialization and deserialization

slobodanadamovic committed:
Prevent creation/updating of API keys with restrictions in mixed cluster

slobodanadamovic committed:
Tests for RoleRestriction

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-for-api-keys

# Conflicts:
#	x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

slobodanadamovic committed:
Revert the order of assertions

slobodanadamovic committed:
Move check back after assertions

slobodanadamovic committed:
Test create/grant/update/bulk_update API keys

elasticsearchmachine commented:
Pinging @elastic/es-security (Team:Security)

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -300,6 +301,7 @@ public void createApiKey(
     ) {
         assert request.getType() != ApiKey.Type.CROSS_CLUSTER || userRoleDescriptors.isEmpty()
             : "owner user role descriptor must be empty for cross-cluster API keys";
+        assert false == hasWorkflowsRestriction(userRoleDescriptors) : "workflows restriction are not supported by user role descriptors";
```
User roles can be resolved by custom role providers. Do you plan to prevent them returning workflow roles? Otherwise, we may need more than an assertion here.

slobodanadamovic replied:
Good question. I think we should prevent it. I will convert this into `IllegalArgumentException`.

ywangd commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java:
```diff
@@ -347,6 +359,10 @@ private static boolean hasRemoteIndices(Collection<RoleDescriptor> roleDescripto
         return roleDescriptors != null && roleDescriptors.stream().anyMatch(RoleDescriptor::hasRemoteIndicesPrivileges);
     }
 
+    private static boolean hasWorkflowsRestriction(Collection<RoleDescriptor> roleDescriptors) {
+        return roleDescriptors != null && roleDescriptors.stream().anyMatch(RoleDescriptor::hasWorkflowsRestriction);
+    }
```
It seems we need a check for: If workflow is used, the API key must have only a single role descriptor.

slobodanadamovic replied:
Yes. I've forgot to add one. I'll fix that and add a test for it.

ywangd reviewed (commented):
This looks promising. I still need to read through the tests. Here are some comments for the main code. Thanks!

slobodanadamovic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java:
```diff
@@ -1517,6 +1559,127 @@ public ApplicationResourcePrivileges build() {
+            return new RoleRestriction(workflows);
+        }
+
+        public static Builder builder() {
+            return new Builder();
+        }
+
+        public static class Builder {
+
+            private String[] workflows;
+
+            public Builder workflows(String... workflows) {
+                this.workflows = workflows;
+                return this;
+            }
+
+            public RoleRestriction build() {
+                return new RoleRestriction(workflows);
+            }
+        }
```
Agreed. I will remove it.

slobodanadamovic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java:
```diff
@@ -1517,6 +1559,127 @@ public ApplicationResourcePrivileges build() {
+            return workflowsRestriction.length > 0;
+        }
+
+        public String[] getWorkflows() {
+            return workflowsRestriction;
```
I went from not having it nested under `Restriction` class and the name slipped the refactoring. I will rename it to `workflows`.

slobodanadamovic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java:
```diff
@@ -1517,6 +1559,127 @@ public ApplicationResourcePrivileges build() {
+
+        private final String[] workflowsRestriction;
+
+        public RoleRestriction(@Nullable String[] workflowsRestriction) {
+            this.workflowsRestriction = workflowsRestriction == null ? Strings.EMPTY_ARRAY : workflowsRestriction;
+        }
```
I like the leniency, but I agree it does not make much sense. I will adjust the parsing and add validation that if  a `restriction` object is provided it must include non-empty `workflows` array. 

This also implies that if a user wants to update an API key and remove workflows restriction it should simply submit update request without `restriction` section. Not saying this is bad. Just wanted to note it down.

e.g.

Create with restriction:
```
"my_restricted_role": {
  "indices": [
    {
      "names": ["books", "movies"],
      "privileges": ["read"]
    }
  ],
  "restriction": {
    "workflows": ["my_workflow"]
  }
}
```
Update to remove restriction:
```
"my_restricted_role": {
  "indices": [
    {
      "names": ["books", "movies"],
      "privileges": ["read"]
    }
  ]
}
```

ywangd replied:
> This also implies that if a user wants to update an API key and remove workflows restriction it should simply submit update request without restriction section. Not saying this is bad. Just wanted to note it down.

I think that's OK. Role descriptors are updated in its entirety. There is no need to explicity say `restriction: {}` or `restriction: null` to remove it. This is just like when you create the API key, you don't need specifiy `restriction: {}` (or `null`) to mean it has no restriction.

 > I'm wondering if there is a convention/reason (from clients perspective) to allow this for workflows as well?

Maybe. But if there is, it is inconsistent. For example, we do not allow null or empty `names` or `privileges` under `indices`. The `field_security` field cannot be null or empty either. If you do prefer it to be a bit more flexible, I can see myself agree with allowing `restriction: {}`. But not `restriction: { workflow: [] }` because it is rather confusing, does it mean no restriction or restricted to be nothing is allowed?

slobodanadamovic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java:
```diff
@@ -1517,6 +1559,127 @@ public ApplicationResourcePrivileges build() {
+        }
+
+        @Override
+        public void writeTo(StreamOutput out) throws IOException {
+            if (out.getTransportVersion().onOrAfter(WORKFLOWS_RESTRICTION_VERSION)) {
+                out.writeStringArray(workflowsRestriction);
+            }
+        }
+
+        public static RoleRestriction readFrom(StreamInput in) throws IOException {
+            if (in.getTransportVersion().onOrAfter(WORKFLOWS_RESTRICTION_VERSION)) {
+                return new RoleRestriction(in);
+            } else {
+                return RoleRestriction.NONE;
+            }
+        }
```
The reason why I moved it inside was to simplify some testing, but I agree that the owner class should have this logic. I'll change it.

slobodanadamovic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java:
```diff
@@ -1517,6 +1559,127 @@ public ApplicationResourcePrivileges build() {
         }
     }
 
+    public static class RoleRestriction implements Writeable, ToXContentObject {
```
That's a good point. I'll remove the prefix. 👍

slobodanadamovic commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/RoleDescriptor.java:
```diff
@@ -1517,6 +1559,127 @@ public ApplicationResourcePrivileges build() {
 
+    public static class RoleRestriction implements Writeable, ToXContentObject {
+
+        public static final TransportVersion WORKFLOWS_RESTRICTION_VERSION = TransportVersion.V_8_9_0;
```
🤦 Forgot about that. Thanks for catching it.

slobodanadamovic committed:
Apply review feedback

- add new transport version
- rename RoleRestriction to Restriction
- move version check logic to RoleDescriptor
- reject null or empty workflows requests
- remove builder class

slobodanadamovic committed:
Validate workflows restriction.

slobodanadamovic committed:
Move transport version logic up to role descriptor class

slobodanadamovic committed:
Use uppercase 'L'.

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-for-api-keys

# Conflicts:
#	x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/ApiKeyService.java

slobodanadamovic committed:
Fix failing tests.

slobodanadamovic committed:
Validate owner user role descriptors

slobodanadamovic committed:
Add null check.

slobodanadamovic commented:
@ywangd This is ready for another round. Thanks for catching mistakes and suggestions.

ywangd commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java:
```diff
@@ -2664,7 +2664,7 @@ private List<RoleDescriptor> randomRoleDescriptors() {
                 randomValueOtherThanMany(
                     rd -> RoleDescriptorRequestValidator.validate(rd) != null,
-                    () -> RoleDescriptorTests.randomRoleDescriptor(false, allowRemoteIndices)
+                    () -> RoleDescriptorTests.randomRoleDescriptor(false, allowRemoteIndices, false)
```
Any reason why we cannot use `randomBoolean()` here?

ywangd commented on x-pack/plugin/security/src/test/java/org/elasticsearch/xpack/security/authc/ApiKeyServiceTests.java:
```diff
@@ -2490,6 +2522,114 @@ public void testValidateApiKeyTypeAndExpiration() throws IOException {
+        );
+
+        final AbstractCreateApiKeyRequest createRequest = mock(AbstractCreateApiKeyRequest.class);
+        when(createRequest.getType()).thenReturn(ApiKey.Type.REST);
```
Nit: 
```suggestion
        when(createRequest.getType()).thenReturn(randomFrom(ApiKey.Type.values()));
```

slobodanadamovic replied:
I have to stick with `ApiKey.Type.REST`, otherwise I trigger cross cluster version validation.

slobodanadamovic replied:
As discussed in person, I'll address handling of Cross Cluster API keys in a followup PRs, depending on the decision we make.

ywangd reviewed (approved):
LGTM

A bunch of minor comments. None is worth another round. Thanks!

Just to summarize what we synced about null and empty value for the new restriction section here: 

---
We definitely should allow the `restriction` field to be absent. I think this is *not* controversial at all.

 ---
I have rather strong preference on *not* allowing empty array for `workflows`. That is, the following should **not** be allowed:
```
{
  "restriction": { "workflows": [] }
}
```

---
I prefer we allow the following empty restriction
```
{
  "restriction": {  }
}
```

 ---
I am netural on the following formats and I'll leave it to you for final decision:
```
{
  "restriction": null
}
```
```
{
  "restriction": { "workflow": null }
}
```

slobodanadamovic committed:
Move TransportVersion constant to RoleDescriptor class and tweak parsing

slobodanadamovic committed:
Use assertOKAndCreateObjectPath

slobodanadamovic committed:
Simplify assertions

slobodanadamovic committed:
Apply review feedback to change error message and inline method

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-for-api-keys

slobodanadamovic committed:
Read and write optionally since workflows can be null

slobodanadamovic commented:
@elasticmachine run elasticsearch-ci/docs

slobodanadamovic commented:
`elasticsearch-ci/docs` failures are unrelated.

slobodanadamovic committed:
Merge branch 'main' of github.com:elastic/elasticsearch into sa-workflows-for-api-keys
